### DISCUSSION TITLE
Typescript types for templater tp object

### DISCUSSION THREAD

**Discussion by metruzanca** [SilentVoid13/Templater#765] [Help]
I'm currently using typescript for my user scripts (compiling with just `tsc *.ts`).
I have obsidian.d.ts (submoduled from obsidian api repo) but would like to also have a file for templater.

Would it be possible to have something in the templater repo that I can submodule into my vault for use with my scripts?

**Comment by vemundrefnin** [comment]
I would also like to have this

**Reply by AB1908** [reply]
I actually have this. I will paste it here, I can't upload into the repo.

**Reply by pentop** [reply]
Hi @AB1908 did you post the types somewhere?

**Reply by AB1908** [reply]
I screwed up, unfortunately. I think @mProjectsCode of Metabind fame (with many other plugins!) may have one. I have one on my PC but it is currently out of commision for the next couple of months. Sorry everyone!

**Reply by mProjectsCode** [reply]
I had one but it's super outdated. But you can use the TS compiler to generate these type def files from the templater source.

**Comment by jameswilliamknight** [comment]
Here's an AI generated definitions file I prompted for while working in a makeshift TS project: [templater.d.ts](https://gist.github.com/jameswilliamknight/2501f4eb9a0e2630be3e835e7f5ab5ac)

YMMV

_if anyone could help me back, I'm looking to set up a modern typescript project for producing `.js` output for Templater, and I'm looking for good examples. Currently I'm just copying versioned .js files from a git repo to my obsidian vault with a bash script._

**Reply by mProjectsCode** [reply]
Why use AI when you can use `tsc` to get accurate and faster results?

Regarding generating `.js` files, why not simply use Templaters build scripts? Why do you need your own setup for this?

**Reply by jameswilliamknight** [reply]
Because I don't know any better. It was describing the usages as I was writing code with `tp`, guess I just got caught up in the moment... I'll look into that.

What is a templater build script? I'm talking about finding a way to manage my templater user scripts in typescript land, any help would be appreciated

**Comment by blakeNaccarato** [comment]
I went ahead and forked Templater over at https://github.com/blakeNaccarato/Templater, modified `package.json`, `tsconfig.json`, and `.gitignore` to get it successfully building declarations with `tsc`, and checked in the built distribution `dist` to source control, so you should be able to get the types for your own template scripting purposes by running e.g

```Shell
npm install --save-dev 'git://github.com/blakeNaccarato/Templater.git#f768a77d02c50ad0d2029aaad112a8b8910584cd'
```

Then in scripts, you can import e.g. `TemplaterPlugin` like this...

#### `getSelectionOrClipboard.ts` (importing from `types.ts` described in follow-up comment, [source](https://github.com/blakeNaccarato/notes/blob/main/templates/getSelOrClip.ts))

```TypeScript
import type TemplaterPlugin from "templater-obsidian";
import type { Templater } from "./types";

/**
 * Get selection, falling back to clipboard.
 */
export default async (): Promise<string> => {
  const tp = (
    (app.plugins.getPlugin("templater-obsidian") as TemplaterPlugin)
      .templater as Templater
  ).current_functions_object;
  return tp.file.selection() || (await navigator.clipboard.readText());
};
```

**Reply by blakeNaccarato** [reply]
Okay, jumping off from my previous comment, I'm working on informing my [`Templater` interface](https://github.com/blakeNaccarato/notes/blob/main/templates/types.ts) (reproduced below) directly from the signatures of e.g. the `generate_selection` method in e.g. `core/functions/internal_functions/file/InternalModuleFile:InternalModuleFile`. See a skeleton of this below with types completed for `system` and partially for `file`. A few things:

- I need to clean up the imports but am not totally clear on how to reduce the verbosity of those imports e.g. `templater-obsidian/dist/core/functions/internal_functions/file/InternalModuleFile`. I'm just picking up TypeScript as I go along.
- Maybe there is some TypeScript magic to reduce the hard-coding of e.g. `SelectionGenerator` and `Selection` types for each method, but hardcoding this isn't so bad really.
- An approach like this could also be fed back into the documentation, automatically keeping documented signature, e.g. for [tp.system.prompt](https://silentvoid13.github.io/Templater/internal-functions/internal-modules/system-module.html#tpsystempromptprompt_text-string-default_value-string-throw_on_cancel-boolean--false-multiline-boolean--false), which looks hardcoded and manually updated in docs right now maybe?

The upsides are that the runtime generated methods can then directly inform the types of the documented user interface, so maybe it's worthwhile to furnish these types upstream to Templater proper. The work would probably look like:

- Expose types via a standalone types module available on NPM or even just this repo. Base types on the [`InternalModule` extensions, e.g. `InternalModuleFile`](https://github.com/SilentVoid13/Templater/blob/4ab3bd5016ead499fbb878bb1b28342789250453/src/core/functions/internal_functions/file/InternalModuleFile.ts#L25), and capturing the signatures of `this.generate_*` included in the `create_static_templates` and `create_dynamic_templates` functions.
- Furnish composable interfaces to statically type the `static_templates`. Compose a default interface with the broadest possible definitions for `dynamic_templates`.
- Still allow extension to narrow the keys associated with `dynamic_templates` (e.g. `tp.file.tags` and `tp.frontmatter`). This could be done by the user via some [fancy typing](https://stackoverflow.com/a/67408203) to "override" generic e.g. `Record<string, string>` types for certain template functions.
- Use this same machinery to drive the written documented interface.

```TypeScript
import type { InternalModuleFile } from "templater-obsidian/dist/core/functions/internal_functions/file/InternalModuleFile";
import type { InternalModuleSystem } from "templater-obsidian/dist/core/functions/internal_functions/system/InternalModuleSystem";

/**
 * Templater interface.
 */
export interface Templater {
  current_functions_object: {
    user: {
      cite: () => Promise<string>;
      getSelOrClip: () => Promise<string>;
      getTimeTrackingEntry: () => Promise<void>;
    };
    file: {
      selection: Selection;
    };
    system: {
      clipboard: Clipboard;
      prompt: Prompt;
      suggester: Suggester;
    };
  };
}

type SelectionGenerator = ReturnType<
  InstanceType<typeof InternalModuleFile>["generate_selection"]
>;
type Selection = (
  ...args: Partial<Parameters<SelectionGenerator>>
) => ReturnType<SelectionGenerator>;

type PromptGenerator = ReturnType<
  InstanceType<typeof InternalModuleSystem>["generate_prompt"]
>;
type Prompt = (
  ...args: Partial<Parameters<PromptGenerator>>
) => ReturnType<PromptGenerator>;

type ClipboardGenerator = ReturnType<
  InstanceType<typeof InternalModuleSystem>["generate_clipboard"]
>;
type Clipboard = (
  ...args: Partial<Parameters<ClipboardGenerator>>
) => ReturnType<ClipboardGenerator>;

type SuggesterGenerator = ReturnType<
  InstanceType<typeof InternalModuleSystem>["generate_suggester"]
>;
type Suggester = (
  ...args: Partial<Parameters<SuggesterGenerator>>
) => ReturnType<SuggesterGenerator>;
```

**Comment by TheRealWolfick** [comment]
Just to add in here, I have a version of this you, or anyone else, are free to use.

[https://github.com/TheRealWolfick/templater-scripts-types](https://github.com/TheRealWolfick/templater-scripts-types)

\- W
